### PR TITLE
Introduce Product Feature value command

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Introduce `SetProductFeatureValuesCommand` to handle the association between product and feature values. But before that some commands regarding FeatureValue needed to be added because they did not exist yet (they will be useful for FeatureValue form page)
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | CI tests green
| Possible impacts? | ~

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23088)

### CODE REVIEW

File: src/Core/Domain/Feature/CommandHandler/AddFeatureValueHandlerInterface.php
Comment (Progi1984) [inline src/Core/Domain/Feature/CommandHandler/AddFeatureValueHandlerInterface.php]:
```suggestion
```

File: src/Core/Domain/Feature/CommandHandler/EditFeatureValueHandlerInterface.php
Comment (Progi1984) [inline src/Core/Domain/Feature/CommandHandler/EditFeatureValueHandlerInterface.php]:
```suggestion
```

File: src/Core/Domain/Feature/QueryHandler/GetFeatureValueForEditingHandlerInterface.php
Comment (Progi1984) [inline src/Core/Domain/Feature/QueryHandler/GetFeatureValueForEditingHandlerInterface.php]:
```suggestion
```

File: src/Core/Domain/Feature/Query/GetFeatureValueForEditing.php
Comment (Progi1984) [inline src/Core/Domain/Feature/Query/GetFeatureValueForEditing.php]:
```suggestion
    public function getFeatureValueId(): FeatureValueId
```

File: tests/Integration/Behaviour/Features/Context/Domain/FeatureValueFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/FeatureValueFeatureContext.php]:
```suggestion
    public function iShouldGetAnErrorThatFeatureValueIsInvalid(): void
```

File: src/Core/Domain/Product/FeatureValue/QueryHandler/GetProductFeatureValuesHandlerInterface.php
Comment (Progi1984) [inline src/Core/Domain/Product/FeatureValue/QueryHandler/GetProductFeatureValuesHandlerInterface.php]:
```suggestion
```

File: src/Core/Domain/Product/FeatureValue/Query/GetProductFeatureValues.php
Comment (Progi1984) [inline src/Core/Domain/Product/FeatureValue/Query/GetProductFeatureValues.php:27]:
```suggestion
```

File: src/Core/Domain/Product/FeatureValue/CommandHandler/SetProductFeatureValuesHandlerInterface.php
Comment (Progi1984) [inline src/Core/Domain/Product/FeatureValue/CommandHandler/SetProductFeatureValuesHandlerInterface.php]:
```suggestion
```

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Add behat tests for `SetProductFeatureValuesCommand`

Need Unit Tests for `src/Core/Domain/Feature/*.php` & `src/Adapter/Feature/*.php`

**Comment by matks** [comment]
> Add behat tests for `SetProductFeatureValuesCommand`
> 
> Need Unit Tests for `src/Core/Domain/Feature/*.php` & `src/Adapter/Feature/*.php`

Hi @Progi1984 I don't think it's a good idea to say unit test a full directory.

Look at the DTO `PrestaShop\PrestaShop\Core\Domain\Product\FeatureValue\Query\GetProductFeatureValues.php` for example 😉  is there value to test this?

I suggest we follow the guidelines of [Robert C. Martin](https://stackoverflow.com/questions/674408/junit-tests-for-pojos) "Test everything that could possibly break". Testing "bags of data" like DTOs, getters and setters have no value and they will be indirectly tested when used in integration tests or even used in other tests (for example when we test GetProductFeatureQueryHandler we will perform assertions on the return value, which is the DTO above).

Source https://stackoverflow.com/questions/674408/junit-tests-for-pojos

**Comment by Progi1984** [comment]
I suggest a thing : "Test everything, and be sure of nothing" (and think about code coverage).

File: src/Core/Domain/Product/FeatureValue/Command/SetProductFeatureValuesCommand.php
Comment (matks) [inline src/Core/Domain/Product/FeatureValue/Command/SetProductFeatureValuesCommand.php:53]: Should we add a validation step for this array, so we reject arrays badly structured?

File: src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php
Comment (matks) [inline src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php:41]: When we migrate Features BO page, we'll have issues 🤔  because from this BO page you can list, add, edit not only "features" entries but also "feature values" entries (the pre-defined values).

We should try to avoid confusion if possible.
Maybe we can rename (in the code) "feature values" into "features pre-defined values" for the other BO page and keep "feature values" here?

Comment (matks) [reply src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php:41]: Anyway, clearing the confusion around this data structure by using PHPDoc could be useful 👍 

```
/**
 * Handles adding of feature value using legacy model.
 *
 * Feature values are the merchant input configured for a given product for a given feature. Example: "Type of paper: Gloss"
 */
```

File: src/Adapter/Product/FeatureValue/Update/ProductFeatureValueUpdater.php
Comment (matks) [inline src/Adapter/Product/FeatureValue/Update/ProductFeatureValueUpdater.php:211]: Syntax sugar suggestion 🍭 
```suggestion
        if ($productFeatureValue->isCustom()) {
```

File: src/Core/Domain/Feature/Command/AddFeatureValueCommand.php
Comment (matks) [inline src/Core/Domain/Feature/Command/AddFeatureValueCommand.php:34]: I'm sorry I'm confused 😅 

Does this mean from BO product page it is possible to create new predefined feature values? I tried from the product page but I did not find where you can do that.

**Comment by matks** [comment]
> I suggest a thing : "Test everything, and be sure of nothing" (and think about code coverage).

That is exactly what I suggest not to do 😅  .

Code coverage might look like a great quality metrics, but it is actually a trap, see [the statement from Martin Fowler](https://martinfowler.com/bliki/TestCoverage.html) about it.

This is a gray area, and it is very hard to find the right balance between "enough testing" and "too much testing", I agree. Many articles explain this very hard situation
- https://medium.com/better-programming/is-code-coverage-a-useless-metric-bc76e0fde9e
- https://hackernoon.com/is-test-coverage-a-good-metric-for-test-or-code-quality-92fef332c871

**Code coverage is a useful metric**, that is for sure. I will not say it should not be looked at. But focusing and making it as high as possible can lead you to a path where you actually work against the project benefits.

There are many classes where it's hard to find out whether extensive testing, and you'll find yourself hesitating whether it's a good thing or not. For example should you test views? Some people say yes, some people say no. Should you test SQL repository classes? Some people say yes, some people say no.

But on getters, setters and DTO I am quite confident that the overall consensus is not to test them. Some people workaround this issue but writing auto-generated tests (just like an IDE can generate getters/setter).

**Comment by Progi1984** [comment]
@matks I disagree with you on the level of quality that we wait for contributions. Maintainers' code must be perfect and an example for others.

So I continue for the "Requested changes" of this PR.

**Comment by matks** [comment]
> @matks I disagree with you on the level of quality that we wait for contributions. Maintainers' code must be perfect and an example for others.

May I quote the wise words of [Google themselves](https://google.github.io/eng-practices/review/reviewer/standard.html) 😉 :

"**In general, reviewers should favor approving a CL once it is in a state where it definitely improves the overall code health of the system being worked on, even if the CL isn’t perfect.**

That is the senior principle among all of the code review guidelines.

There are limitations to this, of course. For example, if a CL adds a feature that the reviewer doesn’t want in their system, then the reviewer can certainly deny approval even if the code is well-designed.

A key point here is that there is no such thing as “perfect” code—there is only better code. Reviewers should not require the author to polish every tiny piece of a CL before granting approval. Rather, the reviewer should balance out the need to make forward progress compared to the importance of the changes they are suggesting. Instead of seeking perfection, what a reviewer should seek is continuous improvement. **A CL that, as a whole, improves the maintainability, readability, and understandability of the system shouldn’t be delayed for days or weeks because it isn’t “perfect.”**"

**Comment by Progi1984** [comment]
@matks Recursivity => https://github.com/PrestaShop/PrestaShop/pull/23088#issuecomment-771657981

[EDIT] I agree with the fact the CL is not perfect. But in this case, the CL is not tested.

File: src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php
Comment (jolelievre) [reply src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php:41]: I already added a comment in that sense in `AddFeatureValueCommand`

Comment (jolelievre) [reply src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php:41]: So to clarify, by default a FeatureValue is a predefined value They are managed outside of the Product page in a list which is common to all products So a FeatureValue is predefined by default

The commands I added are meant for this purpose, the creation and edition of predefined FeatureValue it's the default behaviour so we shouldn't rename it (though I did add a comment in the `AddFeatureValueCommand` to clarify it a bit)

Comment (jolelievre) [reply src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php:41]: By the way that's the reason why these commands are in their own `Feature` namespace and not in a sub namespace of `Product`

Comment (jolelievre) [reply src/Adapter/Feature/CommandHandler/AddFeatureValueHandler.php:41]: Then in the product page you can dynamically create custom feature values, they will added on the fly on marked as `custom` This is the exception use case not the default one (hence the custom)

These custom feature values are handled by the `SetProductFeatureValuesCommand` and are in the scope of the product, they are attached to this product and won't be usable by any other product That's why when you go in the FeatureValue list page you don't see them, you only see pre-defined values

File: src/Adapter/Product/FeatureValue/Update/ProductFeatureValueUpdater.php
Comment (jolelievre) [reply src/Adapter/Product/FeatureValue/Update/ProductFeatureValueUpdater.php:211]: I removed the `isCustom` in the VO so I prefer to rely on `getLocalizedCustomValues()`

File: src/Core/Domain/Product/FeatureValue/Command/SetProductFeatureValuesCommand.php
Comment (jolelievre) [reply src/Core/Domain/Product/FeatureValue/Command/SetProductFeatureValuesCommand.php:53]: Yep, probably a good idea And for this one I'm ready to add unit tests @Progi1984 ^^

File: src/Core/Domain/Feature/ValueObject/FeatureValueId.php
Comment (PierreRambaud) [inline src/Core/Domain/Feature/ValueObject/FeatureValueId.php]:
```suggestion
            throw new InvalidFeatureValueIdException(sprintf('Invalid feature id %d supplied. Feature id must be positive integer.', $featureValueId));
```

It can only be an int

Comment (jolelievre) [reply src/Core/Domain/Feature/ValueObject/FeatureValueId.php]: Done

**Comment by jolelievre** [comment]
@matks @Progi1984 @PierreRambaud I addressed all your feedback (or answered them Ready for review :)

File: src/Adapter/Feature/Repository/FeatureRepository.php
Comment (PierreRambaud) [inline src/Adapter/Feature/Repository/FeatureRepository.php:50]:
```suggestion
            $featureId->getValue(),
```
Don't cast something already casted.

**Review by PierreRambaud — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thank you @jolelievre

**Pull request merged**
